lib/new_relic/agent/instrumentation/ruby_openai/instrumentation.rb

```diff
@@ -0,0 +1,175 @@
+      event.response_choices_finish_reason = response['choices'][0]['finish_reason']
+    end
+
+    def add_embeddings_response_params(response, event)
+      event.response_model = response['model']
+      event.response_usage_total_tokens = response['usage']['total_tokens']
+      event.response_usage_prompt_tokens = response['usage']['prompt_tokens']
+    end
+
+    def parse_api_key
+      'sk-' + headers['Authorization'][-4..-1]
+    end
+
+    # The customer must call add_custom_attributes with llm.conversation_id
+    # before the transaction starts. Otherwise, the conversation_id will be nil.
+    def conversation_id
+      return @nr_conversation_id if @nr_conversation_id
+
+      @nr_conversation_id ||= NewRelic::Agent::Tracer.current_transaction.attributes.custom_attributes[NewRelic::Agent::Llm::LlmEvent::CUSTOM_ATTRIBUTE_CONVERSATION_ID]
+    end
+
+    def create_chat_completion_messages(parameters, summary_id)
+      (parameters[:messages] || parameters['messages']).map.with_index do |message, index|
+        NewRelic::Agent::Llm::ChatCompletionMessage.new(
+          content: message[:content] || message['content'],
+          role: message[:role] || message['role'],
+          sequence: index,
+          completion_id: summary_id,
+          vendor: VENDOR,
+          is_response: false
+        )
+      end
+    end
+
+    def create_chat_completion_response_messages(response, sequence_origin, summary_id)
+      response['choices'].map.with_index(sequence_origin) do |choice, index|
+        NewRelic::Agent::Llm::ChatCompletionMessage.new(
+          content: choice['message']['content'],
+          role: choice['message']['role'],
+          sequence: index,
+          completion_id: summary_id,
+          vendor: VENDOR,
+          is_response: true
+        )
+      end
+    end
+
+    def update_chat_completion_messages(messages, response, summary)
+      messages += create_chat_completion_response_messages(response, messages.size, summary.id)
+      response_id = response['id'] || NewRelic::Agent::GuidGenerator.generate_guid
+
+      messages.each do |message|
+        # TODO: POST-GA: Add metadata from add_custom_attributes if prefixed with 'llm.', except conversation_id
+        message.id = "#{response_id}-#{message.sequence}"
+        message.conversation_id = conversation_id
+        message.request_id = summary.request_id
+        message.response_model = response['model']
+      end
+    end
+
+    def record_openai_metric
+      NewRelic::Agent.record_metric(nr_supportability_metric, 0.0)
+    end
+
+    def segment_noticed_error?(segment)
+      segment&.instance_variable_get(:@noticed_error)
+    end
+
+    def nr_supportability_metric
+      @nr_supportability_metric ||= "Supportability/Ruby/ML/OpenAI/#{::OpenAI::VERSION}"
+    end
+
+    def finish(segment, event)
+      segment&.finish
+      event&.error = true if segment_noticed_error?(segment)
+      event&.duration = segment&.duration
+      event&.record
```

Comment:
Currently this code always calls `segment_noticed_error?`, even when there isn't an event (the `if` is calculated first, then the `&.`). I'm also wondering if `#record` should ever be called if neither `error=` nor `duration=` were called. Here's an attempt to address both issues:

```suggestion
      return unless segment
      
      segment.finish
      return unless event
      
      event.error = true if segment_noticed_error?(segment)
      event.duration = segment.duration
      event.record
```

Reply:
@fallwith, thank you for these suggestions! I think the intended behavior is a little different than your description. 

We still want to record an event if a segment is not present. The `error` and `duration` attributes are optional, and a valid event can be created without them. 

I'm not sure how likely it is to have an event without a segment, but I'd like to protect for that case regardless since they are instances of different event types that can be sent independently of each other.

Given this additional context, what are your thoughts about the following refactor?
```suggestion
      segment&.finish

      return unless event

      if segment
        event.error = true if segment_noticed_error?(segment)
        event.duration = segment.duration
      end

      event.record
```

Reply:
Given your explanation, this code looks good, yep!